encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContextTest.java

```diff
@@ -0,0 +1,137 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders.json;
+
+import androidx.annotation.NonNull;
+
+import com.google.firebase.decoders.FieldRef;
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.ObjectDecoderContext;
+import com.google.firebase.decoders.Safe;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+import com.google.firebase.decoders.json.JsonDataDecoderBuilderContext;
+import com.google.firebase.encoders.FieldDescriptor;
+import com.google.firebase.encoders.annotations.Encodable;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertThrows;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.robolectric.RobolectricTestRunner;
+/**
+ * TODO: TypeToken support TypeVariable, inorder to support decode nested Generic type
+ * class Foo<T> {
+ *   SubFoo<T>
+ * }
+ */
+@RunWith(RobolectricTestRunner.class)
+public class JsonDataDecoderBuilderContextTest {
+
+  static class Foo<T> {
+    int a;
+    boolean b;
+    T t;
+    SubFoo<String> subFoo;
+
+    public Foo(int a, boolean b, T t, SubFoo<String> subFoo) {
+      this.a = a;
+      this.b = b;
+      this.t = t;
+      this.subFoo = subFoo;
+    }
+  }
+
+  static class SubFoo<T> {
+    int a;
+    boolean b;
+    T t;
+
+    public SubFoo(int a, boolean b, T t) {
+      this.a = a;
+      this.b = b;
+      this.t = t;
+    }
+  }
+
+  static class FooObjectDecoder<T> implements ObjectDecoder<Foo<T>> {
+    @NonNull
+    @Override
+    public TypeCreator<Foo<T>> decode(@NonNull ObjectDecoderContext<Foo<T>> ctx) {
+      FieldRef.Primitive<Integer> aField = ctx.decodeInteger(FieldDescriptor.of("a"));
+      FieldRef.Primitive<Boolean> bField = ctx.decodeBoolean(FieldDescriptor.of("b"));
+      FieldRef.Boxed<T> tField = ctx.decode(FieldDescriptor.of("t"), (TypeToken.ClassToken) ctx.getTypeArguments().at(0));
+      FieldRef.Boxed<SubFoo<String>> subFooField = ctx.decode(FieldDescriptor.of("subFoo"), TypeToken.of(new Safe<SubFoo<String>>(){}));
+
+      return (creationCtx ->
+              new Foo<T>(
+                      creationCtx.getInteger(aField),
+                      creationCtx.getBoolean(bField),
+                      (T) creationCtx.get(tField),
```

Comment:
curious why you had to cast to `(T)` here?

Reply:
since generic type `T` in `Box<T>` is downgrade to `Object`, then `creationCtx.get(tField)` give us an `Object` types, therefore i have to explicitly cast it to `T`. 

I will have more investigations on it to see how can I avoid explicitly casting it.

Reply:
At runtime sure, but it should not matter at compile time where it is `T` which is exactly what `Foo`'s constructor expects

Reply:
Oh, I think I see the problem, you need to remove `<T>` on `CreationContext<T>` and make it just `CreationContext`, without this the `TypeCreator` interface is defined with a raw type, hence all generics inference gets disabled and `.get()` return `Object` instead of `T`.